lib/menubar/utils.lua

```diff
@@ -243,17 +246,49 @@ end
 
 --- Parse a directory with .desktop files recursively.
 -- @tparam string dir The directory.
--- @treturn table Paths of found .desktop files.
-function utils.parse_dir(dir)
-    local programs = {}
-    local files = io.popen('find '.. dir .." -name '*.desktop' 2>/dev/null")
-    for file in files:lines() do
-        local program = utils.parse_desktop_file(file)
-        if program then
-            table.insert(programs, program)
+-- @tparam function callback Will be fired when all the files will be parsed.
+-- @tparam table callback.programs Paths of found .desktop files.
+function utils.parse_dir(dir, callback)
+    gio.Async.start(function()
+        local f = gio.File.new_for_path(dir)
+        -- Except for "NONE" there is also NOFOLLOW_SYMLINKS
+        local enum, err = f:async_enumerate_children("standard::name", gio.FileQueryInfoFlags.NONE)
+        if not enum then
+            debug.print_error(err)
+            return callback(nil)
         end
-    end
-    return programs
+        local subdirs_to_parse = 0
+        local files_per_call = 100 -- Actual value is not that important
+        local programs = {}
+        while true do
+            local list, enum_err = enum:async_next_files(files_per_call, glib.PRIORITY_DEFAULT)
+            if enum_err then
+                debug.print_error(enum_err)
+                return callback(nil)
+            end
+            for _, info in ipairs(list) do
+                local file_type = info:get_file_type()
+                local file_path = f:get_child(info:get_name()):get_path()
+                if file_type == 'REGULAR' then
+                    local program = utils.parse_desktop_file(file_path)
+                    if program then
+                        table.insert(programs, program)
+                    end
+                elseif file_type == 'DIRECTORY' then
+                    subdirs_to_parse = subdirs_to_parse + 1
+                    utils.parse_dir(file_path, function(more_programs)
```

Comment:
Hm, could you restructure this? This `utils.parse_dir` starts another async operation. It would be nice if this were done inside of this async operation instead. Otherwise, we could run out of file descriptors and things will start to fail with "too many open FDs" (I had this yesterday when I wrote something that recursively lists files).

To do this, I would suggest to move the function that you call `gio.Async.start` on into a local function of this file that gets `dir` as an argument (and another argument with the table where entries should be added to).

That way, you can replace `utils.parse_dir` with `gio.Async.start(function() local result = {} your_local_function(dir, result) callback(result) end)()` and this recursive call can be replaced with a plain `your_local_function(file_path, result)`.

Another benefit: I think you no longer have to count `subdirs_to_parse` and the invocation of the callback is a lot simpler, too.